src/refactor/CSPHandler.py

```diff
@@ -0,0 +1,116 @@
+import serial
+from system import SatelliteNodes
+
+def getCSPHandler(addr, interface, device, hmacKey, protocol = None):
+    if protocol is None:
+        return CSPHandler(addr, interface, device, hmacKey)
+    elif protocol == "UHF":
+        return UHF_CSPHandler(addr, interface, device, hmacKey)
+    else:
+        raise ValueError("Protocol {} does not exist for CSP handlers".format(protocol))
+
+class CSPHandler(object):
+    __instance = None
+    def __new__(cls, addr, interface, device, hmacKey):
+        if cls.__instance is None:
+            cls.__instance = super(CSPHandler, cls).__new__(cls)
+        return cls.__instance
+
+    # 'addr' groundstation's address
+    # 'interface' interface to use
+    # 'device' device file interface is to write to
+    # 'hmacKey' key to use for HMAC authentication
+    def __init__(self, addr, interface, device, hmacKey):
+        self.connectionManager = ConnectionManager()
+
+        self.myAddr = addr;
+        self.numberOfBuffers = 100
+        self.bufferSize = 1024 #This is max size of an incoming packet
+        libcsp.init(self.myAddr, 'GroundStation', 'model', '1.2.3', self.numberOfBuffers, self.bufferSize)
+        libcsp.hmac_set_key(hmacKey, len(hmacKey))
+
+        self.interfaceName = ""
+        if interface == 'uart':
+            self.ser = self._uart(device)
+        elif interface == 'sdr':
+            self._sdr(device, libcsp.SDR_UHF_9600_BAUD)
+        elif interface == 'sband':
+            raise NotImplementedError("Sband support not yet implemented")
+    
+        stringBuild = ""
+        for node in SatelliteNodes:
+            stringBuild = stringBuild + " {} {} ".format(node.value, self.interfaceName)
+        libcsp.rtable_load(stringBuild)
+
+        libcsp.route_start_task()
+
+    def send(self, server, port, buf : bytearray):
+        packet = packetUtils.makePacket(buf)
+        conn = self.connectionManager.getConn(server,port)
+        libcsp.send(conn, packet)
+        libcsp.buffer_free(packet)
+
+    def receive(self, server, port, timeout):
+        conn = self.connectionManager.getConn(server,port)
+        packet = libcsp.read(conn, timeout)
+        data = None
+        if packet is None:
+            raise Exception("No packet received after {} seconds".format(timeout // 1000))
+        data = packetUtils.breakPacket(packet)
+        libcsp.buffer_free(packet)
+        return data
+
+    def _uart(self, device):
+        """ initialize uart interface """
+        ser = serial.Serial(device,
+        baudrate=115200,
+        bytesize=8,
+        parity='N',
+        stopbits=2,
+        timeout=1)
+
+        libcsp.kiss_init(device, ser.baudrate, 512, 'uart')
+        self.interfaceName = "uart"
+
+    def _sdr(self, device, uhf_baudrate):
+        """ Initialize SDR interface """
+        libcsp.sdr_init(device, 115200, uhf_baudrate, "UHF")
+        self.interfaceName = "UHF"
+
+
+class UHF_CSPHandler(CSPHandler):
```

Comment:
I think pipe mode enable/disable is independent of CSP. Are there other Endurosat specific commands that could be in this class? I'm thinking the command to set transceiver baud rate is a candidate, i.e. outside CSP but related to CSP communication.

Reply:
This class could definitely be extended to support more endurosat stuff. The whole endurosat thing violates our protocol, so to me that justified a subclass to handle that